src/agent/agent.c

```diff
@@ -1805,47 +1805,59 @@ static bool ensure_orch_address(Agent *agent) {
-                int r = get_address(agent->host, &ip_address);
-                if (r != 0) {
-                        hirte_log_errorf("Failed to get IP address from: '%s'", agent->host);
+        _cleanup_free_ char *resolved_ip_address = NULL;
+        if (!host_is_ipv4 && !host_is_ipv6) {
+                int r = get_address(ip_address, &resolved_ip_address);
+                if (r < 0) {
+                        hirte_log_errorf(
+                                        "Failed to get IP address from host '%s': %s",
+                                        agent->host,
+                                        strerror(-r));
                         return false;
                 }
-                hirte_log_infof("Translated '%s' to '%s'", agent->host, ip_address);
-                copy_str(&agent->host, ip_address);
+                hirte_log_infof("Translated '%s' to '%s'", ip_address, resolved_ip_address);
+                ip_address = resolved_ip_address;
         }
 
-        // IPv4
-        if (isIPv4Addr(agent->host)) {
+        if (host_is_ipv4 || is_ipv4(ip_address)) {
                 struct sockaddr_in host;
                 memset(&host, 0, sizeof(host));
                 host.sin_family = AF_INET;
                 host.sin_port = htons(agent->port);
-                r = inet_pton(AF_INET, agent->host, &host.sin_addr);
+                r = inet_pton(AF_INET, ip_address, &host.sin_addr);
                 if (r < 1) {
-                        hirte_log_errorf("INET4: Invalid host option '%s'", agent->host);
+                        hirte_log_errorf("INET4: Invalid host option '%s'", ip_address);
                         return false;
                 }
-                agent->orch_addr = assemble_tcp_address(&host);
-        } else if (isIPv6Addr(agent->host)) { // IPv6
+                _cleanup_free_ char *orch_addr = assemble_tcp_address(&host);
+                if (orch_addr == NULL) {
+                        return false;
+                }
+                agent_set_orch_address(agent, orch_addr);
+        } else if (host_is_ipv6 || is_ipv6(ip_address)) {
                 struct sockaddr_in6 host6;
                 memset(&host6, 0, sizeof(host6));
                 host6.sin6_family = AF_INET6;
                 host6.sin6_port = htons(agent->port);
-                r = inet_pton(AF_INET6, agent->host, &host6.sin6_addr);
+                r = inet_pton(AF_INET6, ip_address, &host6.sin6_addr);
                 if (r < 1) {
-                        hirte_log_errorf("INET6: Invalid host option '%s'", agent->host);
+                        hirte_log_errorf("INET6: Invalid host option '%s'", ip_address);
+                        return false;
+                }
+                _cleanup_free_ char *orch_addr = assemble_tcp_address_v6(&host6);
+                if (orch_addr == NULL) {
                         return false;
                 }
+                agent_set_orch_address(agent, orch_addr);
         }
```

Comment:
Should add one more else block, to fail if is_ipv4(ip_address) fails. I don't think we should get here, but it would be nice if we logged it.

Reply:
agreed "unknown protocol"

Reply:
Good point! Added.